### ISSUE TITLE
Permanent errors support

### ISSUE THREAD

**Issue by arnaud-lb** [cenkalti/backoff#33]
Hi

Sometimes the retries operation knows that it's not going to succeed, ever. For instance, if `http.NewRequest()` fails because the URL is malformed, or if a server returns a `400 Bad Request`, retrying will probably never succeed.

It would be great if there was a way for the operation to specify that it shouldn't be retried.

I would like to propose an implementation of this that doesn't change the _API_ of the package (no changes, no additions), and keeps the implementation simple.

One possible way of doing this would be to allow the operation to return a particular implementation of _error_. Retry() would stop retrying if the operation returned this error:

```
package backoff

type PermanentError struct {
    Err error
}

func (e *PermanentError) String() {
    e.Err.String()
}

// In Retry():

if err = operation(); err == nil {
	return nil
}

if err, ok := err.(*PermanentError); ok {
    return err.Err
}
```

And in user code:

```
backoff.Retry(func() {
    req, err := http.NewRequest("GET", "http://example com", nil)
    if err != nil {
        return &backoff.PermanentError{err}
    }
}, b)
```

**Comment by arnaud-lb** [comment]
This seems to have been requested a lot before, but I believe that you might like the proposed solution, since it's backwards compatible and simple.

For reference, here are the previous requests: #29, #26, #19, #22.

**Comment by cenkalti** [comment]
Hello @arnaud-lb,

I think your solution is simple and perfect. It solves the problem without breaking backwards compatibility. Would you like to send a PR?

**Comment by cenkalti** [comment]
Fixed by #36

**Comment by arnaud-lb** [comment]
Thanks !
